File: C++/String/28_2nd_most_repeating_string.cpp
/*
Idea:
using map store frequencies of all strings
keeping track of highest & 2nd highest frequency value and string value .
*/


#include<bits/stdc++.h>
using namespace std;



class Solution
{
public:
    string secFrequent(string arr[], int n)
    {

        unordered_map <string, int> freq;

        for (int i = 0;i < n;++i)
            freq[arr[i]]++; // inserting in map 

        string first_max_str = "", second_max_str = "";
        int first_max = INT_MIN, second_max = INT_MIN;


        for (auto& entry : freq)
        {
            // finding max & 2nd max frequency strings
            if (entry.second > first_max)
            {
                second_max_str = first_max_str;
                second_max = first_max;

                first_max_str = entry.first;
                first_max = entry.second;
            }
            else if (entry.second > second_max && entry.first != first_max_str)
            {

                second_max_str = entry.first;
                second_max = entry.second;
            }
        }
        return second_max_str;
    }
};


int main()
{
    int t; cin >> t;
    while (t--)
    {
        int n; cin >> n;
        string arr[n];
        for (int i = 0; i < n; ++i)
            cin >> arr[i];
        Solution ob;
        cout << ob.secFrequent(arr, n) << endl;
    }
}

File: Python/String/28_2nd_most_repeating_string.py
"""
Idea:
using dictionary store frequencies of all strings
keeping track of highest & 2nd highest frequency value and string value .
"""


class Solution:
    def secFrequent(self, arr, n):
        freq = {}
        for i in range(n):
            # inserting in dictionary
            if arr[i] in freq:
                freq[arr[i]] += 1
            else:
                freq[arr[i]] = 1

        max1 = max2 = float("-inf")
        string1 = string2 = ""

        for key, value in freq.items():
            # finding max & 2nd max frequency strings
            if value > max1:
                string2 = string1
                max2 = max1

                max1 = value
                string1 = key

            elif value > max2 and key != string1:
                max2 = value
                string2 = key
        return string2


if __name__ == '__main__':
    t = int(input())
    for _ in range(t):
        n = int(input().strip())
        arr = input().strip().split(" ")
        ob = Solution()
        ans = ob.secFrequent(arr, n)
        print(ans)
